### DISCUSSION TITLE
[Compilers] zksolc "Source code for path not found"

### DISCUSSION THREAD

**Discussion by alexhooketh** [zkSync-Community-Hub/zksync-developers#208] [Compilers]
### Which compiler are you using?

- [X] zksolc
- [ ] zkvyper

### Which version of Solidity or Vyper?

0.8.23+commit.f704f362.Darwin.appleclang

### zksolc/zkvyper version?

EraVM Solidity compiler v1.3.17 (LLVM build bab511e161a8c37e5ca50be5eaa43ebb62ed958e)

### Explain the issue

```bash
Due to specifics of my project I decided to use zksolc compiler directly instead of using hardhat or any other dev tools. I took latest release from zksolc-bin repo.

I want to compile a Solidity contract that uses these imports:

// SPDX-License-Identifier: MIT
pragma solidity ^0.8.19;

import "/contracts/l2/BaseL2OpsManager.sol";
import "/contracts/l2/interfaces/IL2OpsManager.sol";
import "@matterlabs/zksync-contracts/l2/system-contracts/Constants.sol";
import "@matterlabs/zksync-contracts/l2/contracts/vendor/AddressAliasHelper.sol";
```

with these command line arguments: `zksolc --include-path=node_modules --base-path=. samples/zksync/ZksyncManager.sol`

And this is what happens in my terminal:

```
jordan@MacBook-Air-Alex xtra-protocol % zksolc --include-path=node_modules --base-path=. samples/zksync/ZksyncManager.sol
Warning: This declaration shadows an existing declaration.
  --> @matterlabs/zksync-contracts/l2/system-contracts/libraries/Utils.sol:85:9:
   |
85 |         uint256 bytecodeLenInWords = _bytecode.length / 32;
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^
Note: The shadowed declaration is here:
  --> @matterlabs/zksync-contracts/l2/system-contracts/libraries/Utils.sol:43:5:
   |
43 |     function bytecodeLenInWords(bytes32 _bytecodeHash) internal pure returns (uint256 codeLengthInWords) {
   |     ^ (Relevant source part starts here and spans across multiple lines).
   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Source code for path `@matterlabs/zksync-contracts/l2/contracts/vendor/AddressAliasHelper.sol` not found
jordan@MacBook-Air-Alex xtra-protocol %
```

Last error looks really strange - it doesn't have TypeError prefix and generally doesn't seem solidity-ish + zksolc exits instantly after it appears. I thought that this could be related to AddressAliasHelper.sol in the library, so I rewrote my code without using this file, and now I get this error:

```
Source code for path `@matterlabs/zksync-contracts/l2/system-contracts/BootloaderUtilities.sol` not found
jordan@MacBook-Air-Alex xtra-protocol %
```

So I believe it isn't related to that file. @matterlabs/zksync contracts is stored in node_modules folder in my project root (I call zksolc from the root). I also tried to compile it with solc using the same arguments and I get this

```
jordan@MacBook-Air-Alex xtra-protocol % solc --include-path=node_modules --base-path=. samples/zksync/ZksyncManager.sol --bin   
Error: Stack too deep. Try compiling with `--via-ir` (cli) or the equivalent `viaIR: true` (standard JSON) while enabling the optimizer. Otherwise, try removing local variables.
   --> @matterlabs/zksync-contracts/l2/system-contracts/BootloaderUtilities.sol:104:34:
    |
104 |             uint256 listLength = encodedNonce.length +
    |                                  ^^^^^^^^^^^^

jordan@MacBook-Air-Alex xtra-protocol %
```

But this isn't quite related to the problem, I'll leave it there just in case
```


### Link to your repo?

-

**Comment by alexhooketh** [comment]
Well i decided just to implement crutchy communication with hardhat compiler in my tests, so it's not really needed for me anymore but if it's a bug (many people told me the compiler is buggy) it's worth to fix it

**Comment by MexicanAce** [comment]
Hi @alexhooketh , this is a bug that the zkSync Compiler team is investigating.

**Comment by omurovec** [comment]
Also having the same issue with zksolc `zksolc-macosx-arm64-v1.3.16`, solc `v0.8.16`

Created an example repo to reproduce the error: https://github.com/omurovec/zksolc-import-error

Would appreciate some more eyes on this if possible

**Comment by hedgar2017** [comment]
Hey @alexhooketh @omurovec - thanks for the report!

In order to resolve such dependencies, the best way is to use hardhat. `zksolc` does not operate on Solidity source code level so does not resolve imports this way.
We'll check the reason and try to fix it, but most likely we'll deprecate input modes that behave this way.

**Reply by omurovec** [reply]
@hedgar2017 Would this mean depricating all foundry support???

It's also worth noting that this has not been an issue in the past with the same configuration.

Additionally when attempting to compile directly with `zksolc`:
```
`~/.zksync/zksolc-macosx-arm64-v1.3.16 ./src/Contract.sol --solc ~/.svm/0.8.16/solc-0.8.16`
```
I receive:
```
Source code for path `~/import-error-example/lib/prb-math/contracts/PRBMath.sol` not found`
```
when I've verified the file exists at that path

**Reply by hedgar2017** [reply]
> @hedgar2017 Would this mean depricating all foundry support???
> 
> 
> 
> It's also worth noting that this has not been an issue in the past with the same configuration.
> 
> 
> 
> Additionally when attempting to compile directly with `zksolc`:
> 
> ```
> 
> `~/.zksync/zksolc-macosx-arm64-v1.3.16 ./src/Contract.sol --solc ~/.svm/0.8.16/solc-0.8.16`
> 
> ```
> 
> I receive:
> 
> ```
> 
> Source code for path `~/import-error-example/lib/prb-math/contracts/PRBMath.sol` not found`
> 
> ```
> 
> when I've verified the file exists at that path

Afaik, the foundry integration is now under development, and it can use the same interface as hardhat.

We'll see if we can fix this option. Could you describe when and with which versions it worked before?

**Reply by omurovec** [reply]
The last version I had working was from June (just after this PR: https://github.com/matter-labs/foundry-zksync/pull/101)
(with zksolc v1.3.11)

I tested the current zksync-foundry head with zksolc v1.3.11 but ran into the initial problem again so it seems like this was introduced in the zksync-foundry repo at some point

I haven't been able to test out other versions in the middle so I'm not sure exactly when this issue was introduced but hopefully this can provide some context. It would be great to have a focus on foundry as many projects now use it over hardhat and is crucial for many developer flows nowadays.

**Reply by hedgar2017** [reply]
@omurovec thanks for clarification. I'll make sure we sort this out with the Foundry developers after holidays.
Could you clarify how you're using Foundry if you're calling `zksolc` directly in cmd?

**Reply by omurovec** [reply]
Thank you, appreciate it.

For clarification, I was only calling zksolc directly for more information when foundry errored on that call. Normally I would not call it directly

**Reply by MexicanAce** [reply]
@omurovec We've made some significant changes/fixes for zkfoundry in the past 3 weeks, could you try again and see if your issues have improved?

**Reply by omurovec** [reply]
Thank you, yes this is solved now

**Comment by ulerdogan** [comment]
Hello, I am having the same issue while using `zksolc` directly in a container to compile [this repo](https://github.com/getclave/clave-contracts). To reproduce:

In terminal
``` bash
SOLC_VERSION="0.8.17-alpine"

docker create -it --name zksolc --entrypoint ash  ethereum/solc:${SOLC_VERSION}

docker cp clave-contracts zksolc:/clave-contracts

docker start -i zksolc
```
In the container
``` bash
ZKSOLC_VERSION="v1.3.22"

wget https://github.com/matter-labs/zksolc-bin/raw/main/linux-amd64/zksolc-linux-amd64-musl-${ZKSOLC_VERSION} -O /bin/zksolc; chmod +x /bin/zksolc

zksolc -O3 \
    --base-path /clave-contracts \
    --include-path /clave-contracts/node_modules \
    --combined-json abi,bin \
    -o /clave-contracts/build \
    /clave-contracts/contracts/**/*.sol
```

Finally, having this error message with other expected warnings from compiler:
```bash
...

Source code for path `@matterlabs/zksync-contracts/l2/system-contracts/BootloaderUtilities.sol` not found
```

Additionally, if I run `zksolc` in my MACOS device, I receive the same error with a different file path:

``` bash
...

Source code for path `@account-abstraction/contracts/samples/callback/TokenCallbackHandler.sol` not found
```

**Comment by dutterbutter** [comment]
@ulerdogan Can you open a separate discussion for your issue please.

**Comment by MexicanAce** [comment]
Latest release of zksolc has resolved this issue. Other issues in this Discussion have also been addressed in their associated threads.
